src/lua/utils.c

```diff
@@ -348,6 +348,54 @@ luaT_newmodule(struct lua_State *L, const char *modname,
 	return 1;
 }
 
+int
+luaT_setmodule(struct lua_State *L, const char *modname)
+{
+	assert(modname != NULL);
+	say_debug("%s(%s)", __func__, modname);
+	assert(lua_gettop(L) >= 1);
+
+	if (lua_isnil(L, -1)) {
+		say_debug("%s(%s): skip nil module value", __func__, modname);
+		lua_pop(L, 1);
+		return 0;
+	}
+
+	/* Get package.loaded[modname]. */
+	lua_getfield(L, LUA_REGISTRYINDEX, "_LOADED");
+	lua_getfield(L, -1, modname);
+
+	/*
+	 * If the same module value is already assigned, do
+	 * nothing.
+	 */
+	if (lua_rawequal(L, -1, -3)) {
+		say_debug("%s(%s): skip the same module value", __func__,
+			  modname);
+		lua_pop(L, 3);
+		return 0;
+	}
+
+	/*
+	 * Verify that the module is not already registered with
+	 * another value.
+	 */
+	assert(lua_isnil(L, -1));
+	lua_pop(L, 1);
+
+	/* Swap package.loaded and the module table. */
+	lua_insert(L, -2);
```

Comment:
Technically, it's better to use `lua_pushvalue(L, -2)` here to avoid copying of the top Lua stack slots. As a result you need to pop 2 elements instead of 1 (that is simple decrement of `L->top`) at the end.

Reply:
Hm. Okay.

I'm bother about readability a bit, but both variants are more or less same in this aspect.

```diff
diff --git a/src/lua/utils.c b/src/lua/utils.c
index ffbbf293f..8becd8da7 100644
--- a/src/lua/utils.c
+++ b/src/lua/utils.c
@@ -383,14 +383,17 @@ luaT_setmodule(struct lua_State *L, const char *modname)
        assert(lua_isnil(L, -1));
        lua_pop(L, 1);
 
-       /* Swap package.loaded and the module table. */
-       lua_insert(L, -2);
+       /* Stack: module table, package.loaded. */
 
        /* package.loaded[modname] = <module table> */
+       lua_pushvalue(L, -2);
        lua_setfield(L, -2, modname);
 
-       /* Pop package.loaded. */
-       lua_pop(L, 1);
+       /*
+        * Drop all the temporary values and the module table from
+        * the stack.
+        */
+       lua_pop(L, 2);
 
        say_debug("%s(%s): success", __func__, modname);
        return 0;
```

Reply:
It seems, it is resolved.